trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/ExtractedWitnessInvariant.java

```diff
@@ -1,119 +1,85 @@
-	private final boolean mIsBefore;
-	private final boolean mIsAfter;
-	private final boolean mIsAt;
 
-	public ExtractedWitnessInvariant(final String invariant, final Collection<String> nodeLabel, final IASTNode match,
-			final boolean isBefore, final boolean isAfter, final boolean isAt) {
+	public ExtractedWitnessInvariant(final String invariant, final Collection<String> nodeLabel, final IASTNode match) {
 		mInvariant = invariant;
 		mNodeLabels = ImmutableSet.copyOf(nodeLabel);
 		mMatchedAstNode = match;
-		mIsBefore = isBefore;
-		mIsAfter = isAfter;
-		mIsAt = isAt;
 	}
 
 	public String getInvariant() {
 		return mInvariant;
 	}
 
+	@Override
 	public ImmutableSet<String> getNodeLabels() {
 		return mNodeLabels;
 	}
 
-	public int getStartline() {
+	private int getStartline() {
 		return mMatchedAstNode.getFileLocation().getStartingLineNumber();
 	}
 
-	public int getEndline() {
+	private int getEndline() {
 		return mMatchedAstNode.getFileLocation().getEndingLineNumber();
 	}
 
-	public boolean isBefore() {
-		return mIsBefore;
-	}
-
-	public boolean isAfter() {
-		return mIsAfter;
-	}
-
-	public boolean isAt() {
-		return mIsAt;
-	}
-
 	public IASTNode getRelatedAstNode() {
 		return mMatchedAstNode;
 	}
 
-	public ExtractedWitnessInvariant merge(final ExtractedWitnessInvariant other) {
-		if (other == null) {
-			return this;
-		}
-		if (mMatchedAstNode != other.mMatchedAstNode) {
-			throw new IllegalArgumentException("Cannot merge WitnessInvariants that are matched to different nodes");
-		}
-		final StringBuilder newInvariant = new StringBuilder();
-		newInvariant.append('(');
-		newInvariant.append(mInvariant);
-		newInvariant.append(')');
-		newInvariant.append("||");
-		newInvariant.append('(');
-		newInvariant.append(other.mInvariant);
-		newInvariant.append(')');
-
-		final Set<String> newNodeLabels = new HashSet<>();
-		newNodeLabels.addAll(mNodeLabels);
-		newNodeLabels.addAll(other.mNodeLabels);
-
-		return new ExtractedWitnessInvariant(newInvariant.toString(), newNodeLabels, mMatchedAstNode, isBefore(),
-				isAfter(), isAt());
-	}
-
 	@Override
 	public String toString() {
-		return getBAACode() + " [L" + getStartline() + "-L" + getEndline() + "] " + mInvariant;
+		return getLocationDescription() + " [L" + getStartline() + "-L" + getEndline() + "] " + mInvariant;
 	}
 
-	public String toStringWithCNode() {
-		return toString() + " --- " + getRelatedAstNode().getRawSignature();
-	}
-
-	public String toStringWithWitnessNodeLabel() {
-		return getBAACode() + " [L" + getStartline() + "-L" + getEndline() + "]["
-				+ mNodeLabels.stream().collect(Collectors.joining(", ")) + "]" + mInvariant;
+	protected abstract String getLocationDescription();
+
+	protected ExpressionResult instrument(final ILocation loc, final IDispatcher dispatcher) {
+		ACSLNode acslNode = null;
+		try {
+			checkForQuantifiers(mInvariant);
+			acslNode = Parser.parseComment("lstart\n assert " + mInvariant + ";", getStartline(), 1);

```

Comment:
Just as a fun thought: If `mInvariant` is `true;\nwhile(true)`, would we then validate the witness even potentially for an incorrect program?

Reply:
I am not entirely sure, how parsing ACSL comments works, but I think we just crash with invalid syntax in that case.